dvc/api.py

```diff
@@ -7,31 +7,38 @@
 
 
 class UrlNotDvcRepoError(DvcException):
-    """Thrown if given URL is not a DVC repository.
-
-    Args:
-        url (str): URL to the repository
-    """
+    """Thrown if the given URL is not a DVC repository."""
 
     def __init__(self, url):
         super().__init__("'{}' is not a DVC repository.".format(url))
 
 
 def get_url(path, repo=None, rev=None, remote=None):
-    """Returns URL to the storage location of a data artifact tracked
-    by DVC, specified by its path in a repo.
+    """
+    Returns the URL to the storage location of a data file or directory tracked
+    in a DVC repo. Its formed by reading the remote configuration, and the
+    DVC-file where the given path is an output. For Git repos, HEAD is used
+    unless a rev argument is supplied.
+
+    Raises UrlNotDvcRepoError if repo is not a DVC project.
 
-    NOTE: There's no guarantee that the file actually exists in that location.
+    NOTE: This function does not check for the actual existence of the file or
+    directory in the remote storage.
     """
     with _make_repo(repo, rev=rev) as _repo:
-        _require_dvc(_repo)
+        if not isinstance(_repo, Repo):
+            raise UrlNotDvcRepoError(_repo.url)
         out = _repo.find_out_by_relpath(path)
         remote_obj = _repo.cloud.get_remote(remote)
         return str(remote_obj.checksum_to_path_info(out.checksum))
 
 
 def open(path, repo=None, rev=None, remote=None, mode="r", encoding=None):
-    """Context manager to open a tracked file as a file object."""
+    """
+    Open file or model (path) tracked in a repo (by DVC or Git), and and return
```

Comment:
The first sentence is very confusing and misleading. Also hard to read because all of the parentheses. 

> Open file or model (path)

How can I open a model or a path?

> a repo (by DVC or Git)

does it mean "dvc or git repo"?

> return a file object

It doesn't return a file object, it returns a context manager. So this is misleading, which will lead to people trying to use it wrong and frustrating. Maybe the only easy way to explain how it works is a small example.

Reply:
> It doesn't return a file object, it returns a context manager.

this is my bad. I just didn't expect this behavior from the function and didn't believe when I read it in the initial PR for docs. It's a separate question why did we make it this way, but we need to find a way to write it now, just "returns a context manager" even worse than "returns a files object" (taken from stdlib's open, quite excepted to behave this way here I would say. 

@Suor any suggestions?

Reply:
My suggestion above: show a quick example.

Reply:
The reason we made it a context manager is the control of resources though, which is not working very well since we added caching. We might reconsider.

Reply:
> Open file or model (path)

I'm trying to use the name of the parameters in the description, since we don't have structured sections such as `Params:` or `@param path str blah`. I agree the parentheses here were unnecessary: will use "file or model in the supplied path" instead.

p.s. I still think using backquotes would make including parameter names easier in these descriptions e.g. just "``file or model in `path` ``" or "``tried unless `remote` is supplied.``" – so its obvious the literals are arg names, not just words in a longer sentence (rn there's some ambiguity).

> tracked in a repo (by DVC or Git)

The stuff in parentheses refers more to "tracked" than to to "repo". Should we change to "tracked (by DVC or Git) in a repo"? I don't like how the parentheses interrupt the reading there. Thinking... ⏳

Reply:
> tracked in a repo (by DVC or Git)

OK I'm changing this parenthesis to "(both DVC and plain Git repos are supported)". Better?

Reply:
> It doesn't return a file object, it returns a context manager... Maybe the only easy way to explain how it works is a small example.

OK how about I say "and produces a file object (see example below)" and add the example? Done now, please check!

Reply:
I changed "produces" for "generates" now since I see yielding is involved here.

Or should I use "yields" directly, implying it's a generator? (The context manager returned is implemented [as a generator](https://book.pythontips.com/en/latest/context_managers.html#implementing-a-context-manager-as-a-generator), right?)

> See also the latest updates in iterative/dvc.org/pull/908, matching all this.

Reply:
`yield ` is an implementation detail, there are other ways to make a context manager, so we don't have a generator here. Still hard to read, I would simply say:

```
Opens a file at specified path in a given repo. 
Should be used as a context manager:

    with dvc.api.open(
        "path/from/repo/root/file.ext",
        repo="https://example.com/url/to/repo",
    ) as fd:
        # ... Handle file object fd

Tries to stream file contents without downloading it. ...
Work for both dvc managed files and git handled ones, 
including ones in non-dvc git repos.
```

Reply:
OK I'm using something like that now, please check.

> I'm not sure I like that the only mention of the file object generated is in the example code comment (`Handle file object fd`) but OK, sounds like open() may change in the near future anyway (to behave more like builtin open)